Return SQLite status code on error

hsjoberg opened the pull request:
When an error occurs, the library will throw an exception.
However, The Error object that is thrown does not contain the actual status code
from Sqlite. This commit addresses this by extending the Error class to
be able to contain an error code member property.

All Sqlite error codes are also provided in constants.js now.
https://www.sqlite.org/c3ref/c_abort.html

No tests for this yet, I could look into that.

dyedgreen commented:
Hey @hsjoberg thanks for submitting so many PRs! I wasn't really ready for it yet, so I just set up the tests to (hopefully) also run on PRs. Could you push something/ force push something to trigger the tests?

I'll do a more in-depth review later, but could you add some documentation to the new error class? I don't want to export anything from the public interface that is not documented 😄 See the other classes for how to format the comment, but basically it should look like
```javascript
/*
SqliteError

Description
*/
constructor() { ...

/*
SqliteError.statusName

Description
*/
statusName() { ...
```
(where the title is needed for the doc generation script to correctly parse the comments)

Another thing: Could you open issues for this PR (and the bigger PR for named parameters)? (And please, if you have time, note down anything that would have helped you get started, so that I can start writing something like a `CONTRIBUTING.md` 😅)

Thank you very much for submitting the 3 PRs! 😊

hsjoberg commented:
Hi @dyedgreen!

> Could you push something/ force push something to trigger the tests?

Sure!

> I'll do a more in-depth review later, but could you add some documentation to the new error class? I don't want to export anything from the public interface that is not documented 😄

Oh yes, definitely! I'll fix that. ☺️

> Another thing: Could you open issues for this PR (and the bigger PR for named parameters)? (And please, if you have time, note down anything that would have helped you get started, so that I can start writing something like a CONTRIBUTING.md 😅)

Definitely! I'll open issues.

Getting started wasn't very difficult on Linux, as setting up esmcripten was a very nice experience.
The only thing I really got stuck on was the symbol exporting when compiling the binary (well I still don't know why `KEEPALIVE` doesn't work for me).
I'll try to note down some things.

The project itself has a strong foundation so it wasn't really difficult to get familiar. The only thing that bugged me was that it is written in JS instead of TS, but I see that you have that as a Goal.

> Thank you very much for submitting the 3 PRs! 😊

No worries. Thank you for starting the project -- I was thinking about doing a Deno lib for SQLite myself, but being new to Deno I didn't realize that the obivous choice would be to compile SQlite to WebAssembly, really cool!

hsjoberg commented:
@dyedgreen CI got triggered, seems to pass.

hsjoberg commented:
Some basic documentation has been added for `SqliteError`.

`rows.js` now also properly use `SqliteError` with result code.
https://github.com/dyedgreen/deno-sqlite/blob/151618815281ca6778f9c647928b5460d8236357/src/rows.js#L64,L69

dyedgreen commented on src/rows.js:
```diff
@@ -52,17 +53,20 @@ export class Rows {
-        // TODO: Make more helpful
         this.done();
-        throw new Error("Internal error.");
+        throw new SqliteError(
```
This won't work as expected (you need to reference the db's id, not the statement's id). I think the `_error` function of the db object should be used here (it's used in DB itself whenever an error is generated based on a returned code):

```suggestion
        throw this._db._error(status);
```

hsjoberg replied:
My bad! :sweat_smile:
Fixed.

hsjoberg commented:
Thank you for the review @dyedgreen, I will look at it as soon as I have time. :slightly_smiling_face:

hsjoberg commented on src/constants.js:
```diff
@@ -1,11 +1,40 @@
+  sqliteRow:         100, // sqlite3_step() has another row ready
+  sqliteDone:        101, // sqlite3_step() has finished executing
+
+  stmtLimit:         1000,
```
Sure thing, make sense.
To be honest, I don't know what they stand for, I would appreciate help here. :stuck_out_tongue:

dyedgreen replied:
The interface is designed to never leak any memory (with the sole exception of using the exposed `malloc`). To achieve this, open database handles and open transactions are stored in fixed sized registries (hence they all have id's instead of pointers).

`stmtLimit` -> statement limit was reached: the statement registry is full, no more statements can be opened
`noStms` -> registry entry at this id is empty
`databaseLimit` -> database limit was reached: the database registry is full, no more databases can be opened
`noDatabase` -> registry entry at this id is empty

hsjoberg replied:
Thanks for the explanation.

Added comments.

hsjoberg commented on mod.ts:
```diff
@@ -44,4 +45,4 @@ async function save(db: DB, path?: string): Promise<void> {
 }
 
-export { open, save, DB, Empty };
+export { open, save, DB, Empty, constants };
```
Fixed.

hsjoberg commented on src/db.js:
```diff
@@ -75,6 +76,10 @@ export class DB {
    * !> Any returned Rows object needs to be fully
    * iterated over or discarded by calling
    * `.done()`.
+   *
```
Fixed.

hsjoberg commented on src/db.js:
```diff
@@ -187,21 +193,21 @@ export class DB {
         break;
       default:
         // SQLite error
         const msg = `sqlite error: ${getStr(this._wasm, this._wasm.get_sqlite_error_str(this._id))}`;
```
Fixed

hsjoberg commented on src/error.js:
```diff
@@ -0,0 +1,63 @@
+   *
+   */
+
+  name = "SqliteError";
```
Definitely. Done.

hsjoberg commented on src/error.js:
```diff
@@ -0,0 +1,63 @@
+   *
+   * Beyond the SQLite error code, this member
+   * can also contain custom error codes specific
+   * for this library:
```
Thanks!

hsjoberg commented on src/error.js:
```diff
@@ -0,0 +1,63 @@
+   * can also contain custom error codes specific
+   * for this library:
+   *
+   *     stmtLimit:         1000
```
Oh yes a markdown table makes more sense.

I think it would be appropriate to have all codes here, I'm fixing it for now until we decide what to do. It looks pretty bad on the generated docs site though.

dyedgreen replied:
I think leaving them in is better. You can try making the table wider, by having 4 columns, that may look better. But I'd be willing to trade prettiness for completes in the api doc 😄

hsjoberg replied:
Yes, sounds like a good trade-off.

hsjoberg commented on src/error.js:
```diff
@@ -0,0 +1,63 @@
+   *     noStmt:            1001
+   *     databaseLimit:     1002
+   *     noDatabase:        1003
+   *
```
Fixed

hsjoberg commented on src/error.js:
```diff
@@ -0,0 +1,63 @@
+  /**
+   * SqliteError.codename
+   *
+   * String representation
```
Thanks! Fixed

hsjoberg commented on src/error.js:
```diff
@@ -0,0 +1,63 @@
+   * to also contain class members for error code
+   * and error code name
+   *
+   *     try {
```
Makes sense, fixed.

hsjoberg commented on src/error.js:
```diff
@@ -0,0 +1,63 @@
+   * see the SQLite docs for more
+   * information about each error code.
+   *
+   * [https://www.sqlite.org/c3ref/c_abort.html](https://www.sqlite.org/c3ref/c_abort.html)
```
Fixed.

hsjoberg commented:
Thanks again for the review @dyedgreen, I've addressed most of the points.

Other than addressing the feedback, I've changed `constants.js` to export in a more ES6-module conventional way -- instead of having a default export containing an object containing each thing, the file now exports each thing separately, which makes it easier to import the specific thing you want (see `mod.ts` for example). 

I've made a new commit to make it more clear what has been changed, I'll squash once we're ready to merge.

dyedgreen commented on docs/examples.md:
````diff
@@ -55,6 +55,20 @@ db.query("INSERT INTO people (name, email) VALUES (?, ?)", name, email);
+
+```javascript
+try {
+  db.query("SELEC * FROM people");
````
Maybe add a comment, or make the syntax error more obvious
```suggestion
  // Syntax error
  db.query("SELEC * FROM people");
```
or
```suggestion
  db.query("NOT A QUERY");
```

hsjoberg replied:
Fixed.

dyedgreen commented on src/error.js:
```diff
@@ -0,0 +1,89 @@
+   * `codename` would be sqliteConstraint.
+   *
+   */
+  get codename() {
```
Sorry, this slipped in the last review, but can we stick with the naming convention from deno_std:
```suggestion
  get codeName() {
```

(also remember to update the example in the docs, and the key in the documenting comment)

hsjoberg replied:
Fixed.

dyedgreen commented on src/rows.js:
```diff
@@ -1,5 +1,6 @@
 import { getStr } from "./wasm.js";
-import constants from "./constants.js";
+import * as constants from "./constants.js";
+import SqliteError from "./error.js";
```
Remove unused import

hsjoberg replied:
Oops! Fixed.

dyedgreen commented on src/db.js:
```diff
@@ -75,12 +76,13 @@ export class DB {
    * !> Any returned Rows object needs to be fully
    * iterated over or discarded by calling
    * `.done()`.
+   *
```
Can you remove the extra newline in the comment? (This looks nicer, especially in the generated docs 😅)

hsjoberg replied:
Fixed.

dyedgreen commented on src/error.js:
```diff
@@ -0,0 +1,89 @@
+   * Extension over the standard JS Error object
+   * to also contain class members for error code
+   * and error code name.
+   *
```
Can you remove the extra newline?

hsjoberg replied:
Fixed.

dyedgreen commented on src/error.js:
```diff
@@ -0,0 +1,89 @@
+   *
+   * These codes are accessible via
+   * the exported `status` object.
+   *
```
Can you remove the extra newline?

hsjoberg replied:
Fixed.

dyedgreen commented on src/error.js:
```diff
@@ -0,0 +1,89 @@
+   *
+   * For example, if `code` is 19,
+   * `codename` would be sqliteConstraint.
+   *
```
Can you remove the extra newline?

hsjoberg replied:
Fixed.

hsjoberg commented:
@dyedgreen PTAL. Addressed all of the points. :slightly_smiling_face: 

Made another commit so it is clearer what changed, will squash once we're ready.
I also committed api.md, I missed that I was supposed to be committed as well.

dyedgreen commented:
> @dyedgreen PTAL. Addressed all of the points. 🙂
> 
> Made another commit so it is clearer what changed, will squash once we're ready.
> I also committed api.md, I missed that it was supposed to be committed as well.

Thanks! I'll have a look later, when I have time 😃

dyedgreen commented on docs/api.md:
````diff
@@ -8,7 +8,7 @@ rerun the generator, to avoid loosing the changes.
 ## How to import
 ```javascript
-import { open, save, DB, Empty } from "https://deno.land/x/sqlite/mod.ts"
+import { open, save, DB, SqliteError, Empty } from "https://deno.land/x/sqlite/mod.ts"
````
This is not your fault, but there was a bug in the doc generator which was corrected after your branch. Could you rebase on master to get the fix and rerun the generator? Apologies for this inconvenience 😅

hsjoberg replied:
No problem. :slightly_smiling_face:  Okay I rebased from master and reran the generator.

hsjoberg commented:
@dyedgreen Okay, AFAICT everything should be ready to go. I've squashed everything to one commit.
 
> thank you very much for the PR!

You're welcome. :slightly_smiling_face:
